### PR TITLE
Enable torch_speed_benchmark to accept different memory formats.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36202 Enable torch_speed_benchmark to accept different memory formats.**
* #35791 Enable stateless XNNPACK linear.
* #35790 Enable stateless XNNPACK convolutions.
* #36032 Mobile CPU allocator.

Differential Revision: [D20970216](https://our.internmc.facebook.com/intern/diff/D20970216)

### CODE REVIEW

File: binaries/speed_benchmark_torch.cc [hunk 1]
```diff
@@ -37,12 +37,13 @@ C10_DEFINE_string(
     "semicolon to separate the dimension of different "
     "tensors.");
 C10_DEFINE_string(input_type, "", "Input type (uint8_t/float)");
+C10_DEFINE_string(input_format, "", "Input memory format (NHWC/NCHW)");
```
Comment (AshkanAliabadi) [inline binaries/speed_benchmark_torch.cc:40]: This parameter not having a default value means AIBench invocations must change or they will error.  I think it's better to make this explicit (hence why I opted for not having a default value.)

Comment (kimishpatel) [reply binaries/speed_benchmark_torch.cc:40]: There are too many places this will require update, especially in AIBench. Plus it also introduce discrepancy between this binary and our internal binary that depends on lite predictor. I strongly suggest we keep the default value of NCHW. Otherwise all full JIT runs will be broken.

File: binaries/speed_benchmark_torch.cc [hunk 2]
```diff
@@ -83,10 +85,15 @@ int main(int argc, char** argv) {
 
   std::vector<std::string> input_dims_list = split(';', FLAGS_input_dims);
   std::vector<std::string> input_type_list = split(';', FLAGS_input_type);
+  std::vector<std::string> input_format_list = split(';', FLAGS_input_format);
```
Comment (kimishpatel) [inline binaries/speed_benchmark_torch.cc:88]: Would it make sense to call this arg `input_memory_format` instead of `input_format`? I know `input_memory_format` is mouthful but I thought it is more explicit. Just a suggestion.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Generally looks fine with me, but I left a comment regarding default format that you pointed out.

File: binaries/speed_benchmark_torch.cc (hunk 2 above)
Comment (AshkanAliabadi) [reply binaries/speed_benchmark_torch.cc:88]: Thanks, will change.

File: binaries/speed_benchmark_torch.cc (hunk 1 above)
Comment (AshkanAliabadi) [reply binaries/speed_benchmark_torch.cc:40]: Thanks, will default to NHWC.

Comment (kimishpatel) [reply binaries/speed_benchmark_torch.cc:40]: Note that, that format is valid only for tensors with dims=4. So it wont work for all cases we have in benchmarks. Probably best to do NCHW.

Comment (dreiss) [reply binaries/speed_benchmark_torch.cc:40]: I agree to default to ContiguousFormat.  Also, I think the names should be changed to contiguous_format and channels_last, since that's the terminology that PyTorch uses.

File: binaries/speed_benchmark_torch.cc [hunk 3]
```diff
@@ -37,12 +37,13 @@ C10_DEFINE_string(
...
 C10_DEFINE_string(input_type, "", "Input type (uint8_t/float)");
+C10_DEFINE_string(input_format, "", "Input memory format (NHWC/NCHW)");
 C10_DEFINE_bool(
   print_output,
   false,
   "Whether to print output with all one input tensor.");
-C10_DEFINE_int(warmup, 0, "The number of iterations to warm up.");
-C10_DEFINE_int(iter, 10, "The number of iterations to run.");
+C10_DEFINE_int(warmup, 10, "The number of iterations to warm up.");
+C10_DEFINE_int(iter, 100, "The number of iterations to run.");
```
Comment (dreiss) [inline binaries/speed_benchmark_torch.cc:46]: This should not be part of this change.

**Review by dreiss — CHANGES_REQUESTED** [review]

File: binaries/speed_benchmark_torch.cc (hunk 3 above)
Comment (AshkanAliabadi) [reply binaries/speed_benchmark_torch.cc:46]: Will remove.

File: binaries/speed_benchmark_torch.cc (hunk 2 above)
Comment (AshkanAliabadi) [reply binaries/speed_benchmark_torch.cc]: Done.

File: binaries/speed_benchmark_torch.cc (hunk 3 above)
Comment (AshkanAliabadi) [reply binaries/speed_benchmark_torch.cc]: Done

File: binaries/speed_benchmark_torch.cc (hunk 1 above)
Comment (AshkanAliabadi) [reply binaries/speed_benchmark_torch.cc]: Done

File: binaries/speed_benchmark_torch.cc [hunk 4]
```diff
@@ -77,10 +81,17 @@ std::vector<c10::IValue> create_inputs() {
...
+      split(';', FLAGS_input_memory_format);
+
   CAFFE_ENFORCE_EQ(
       input_dims_list.size(),
       input_type_list.size(),
       "Input dims and type should have the same number of items.");
+  CAFFE_ENFORCE_EQ(
+      input_dims_list.size(),
+      input_memory_format_list.size(),
+      "Input dims and format should have the same number of items.");
```
Comment (dreiss) [inline binaries/speed_benchmark_torch.cc:94]: I'm a little concerned that this will break existing use cases where we're explicitly specifying the dims and types.  If there are more than two inputs, this will cause the benchmark to fail if the memory format isn't specified, right?

**Review by dreiss — APPROVED** [review]
